---
title: "A novel role for alkyladenine DNA glycosylase in regulating alkylation-induced ER stress"
subtitle: "Supplemental Data 2"
author: "L Milano, CF Charlier, R Andreguetti, A Aljohani, E Healing, MP Thom√©, R Elliott, LD Samson, G Lenz, JAP Henriques, A Nohturfft and LB Meira"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
    toc_depth: 3
    code_folding: "hide"
    number_sections: FALSE
    theme: "readable"
    highlight: "tango"
    fig_caption: TRUE
    css: "./source/styles.css"
---

```{js}
function myFunction(id) {
  var x = document.getElementById(id);
  if (x.style.display === 'none') {
    x.style.display = 'block';
  } else {
    x.style.display = 'none';
  }
}
```

<div class='comments'>
*Note*: This analysis requires input data generated by an earlier script.
</div>

## Setup  

* Clear memory and console  
* Load packages  
* Get current script name  

```{r CLEAR MEMORY AND PACKAGES AND CONSOLE, results="hide"}
# Clear memory
rm(list=setdiff(ls(all=TRUE), c(".Random.seed")))
# Clear console
cat("\014")
```

## Set up  
```{r Packages, include=FALSE, results="hide"}
library(dplyr)
library(magrittr)
library(openxlsx)
library(enrichR)
```


```{r Get script name,  class.output="txt_output"}
this.script <- rstudioapi::getActiveDocumentContext()$path %>% basename
cat("Script:", this.script)
```



## Load data
### Load differentially expressed genes  
```{r Read DE genes, class.output="txt_output"}
# list.files("data")
diff.genes.file <- "./data/02_Suppl_Table_01.Rmd.DEgenes.RDS"
stopifnot(file.exists(diff.genes.file))
test.results.genes <- readRDS(diff.genes.file)
cat("File read:", diff.genes.file)
rm(diff.genes.file)
```

```{r results="asis"}
x <- paste(utils::capture.output(str(test.results.genes)), collapse="<br>\n", sep="")
details::details(x, summary="Show summary of DE genes", lang=NULL)
rm(x)
```

<div class='comments'>
The following count values will differ slightly from those in Figure 1 (A and B) because those in Figure 1 represent microarray probes; however, the values shown below are based on gene symbols, and many probes map to multiple gene symbols.
</div>

### Show differentially regulated gene counts  
```{r}
lapply(test.results.genes, function(tr) {
  sapply(tr, length) %>% as.matrix
}) %>% do.call(cbind, .) %>% 
  set_colnames(names(test.results.genes))
```
```{r}
g.lists <- lapply(seq_along(test.results.genes), function(i) {
  name.i <- names(test.results.genes)[i]
  list.i <- test.results.genes[[i]][-3]
  names(list.i) <- paste(name.i, names(list.i), sep="_")
  list.i
}) %>% purrr::flatten(.)
```


```{r results="asis"}
x <- paste(utils::capture.output(str(g.lists)), collapse="<br>\n", sep="")
details::details(x, summary="Show summary of gene list object", lang=NULL)
rm(x)
```


## Gene set enrichment using enrichR  
### Check available libraries  
```{r class.output="txt_output"}
enrichr.libs.df <- listEnrichrDbs()
lib.count <- nrow(enrichr.libs.df)
cat("Number of available enrichr libraries:", lib.count)
```


```{r}
# help(package="enrichR")
enrichr.libs.df %>% 
  dplyr::select(libraryName, numTerms) %>% 
  DT::datatable(., escape=FALSE, rownames = FALSE, caption="All enrichr libraries:")
```

### Select subset of enrichr libraries to search  
```{r}
libs.omit <- c("BioCarta_2013", "BioCarta_2015", "ChEA_2013", "ChEA_2015",
               "ENCODE_Histone_Modifications_2013", "ENCODE_TF_ChIP-seq_2014",
               "Enrichr_Libraries_Most_Popular_Genes", "Enrichr_Users_Contributed_Lists_2020",
               "Genes_Associated_with_NIH_Grants", "GO_Biological_Process_2013",
               "GO_Biological_Process_2017", "GO_Biological_Process_2017b", "GO_Biological_Process_2018",
               "GO_Cellular_Component_2013", "GO_Cellular_Component_2017", "GO_Cellular_Component_2017b",
               "GO_Cellular_Component_2018", "GO_Molecular_Function_2013", "GO_Molecular_Function_2017",
               "GO_Molecular_Function_2017b", "GO_Molecular_Function_2018", "HumanCyc_2015",
               "KEA_2013", "KEGG_2013", "KEGG_2015", "KEGG_2019_Human", "NCI-Nature_2015",
               "NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions",
               "NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions",
               "NIH_Funded_PIs_2017_Human_AutoRIF", "NIH_Funded_PIs_2017_Human_GeneRIF",
               "Panther_2015", "Rare_Diseases_AutoRIF_ARCHS4_Predictions",
               "Rare_Diseases_AutoRIF_Gene_Lists", "Reactome_2013", "Reactome_2015",
               "SysMyo_Muscle_Gene_Sets", "WikiPathways_2013", "WikiPathways_2015",
               "WikiPathways_2019_Human"
               )
```


```{r class.output="txt_output"}
libs.select.df <- enrichr.libs.df %>% 
  dplyr::filter(! libraryName %in% libs.omit)
cat("Number of libraries selected:", nrow(libs.select.df))
```


```{r class.output="txt_output"}
geneset.count <- sum(libs.select.df$numTerms)
cat("Total number of library genesets:", scales::comma(geneset.count))
```



```{r}
colms <- c("Library", "Term", "Overlap", "P.value", "Adjusted.P.value", "Combined.Score", "Genes")
dummy.df <- as.data.frame(t(matrix(rep(NA, length(colms)))),
                          stringsAsFactors=FALSE) %>% 
  set_colnames(colms)
dummy.df
```


### Perform enrichment analysis  
```{r class.output="txt_output"}
cutoff.adj.p <- 1e-06
dbs <- libs.select.df$libraryName
# dbs <- c("GO_Molecular_Function_2018", "GO_Cellular_Component_2018",
#          "GO_Biological_Process_2018")
cat("Adjusted p-value cutoff:", cutoff.adj.p)
```


```{r}
# The enrichr() function generates a list of data frames.
# The length of the list corresponds to the length of dbs.
# List names are dbs values (Library names).

do.enrichr <- function(gs, dbs, cutoff.p) {
  a <- enrichr(gs, dbs) # generates list of data frames
  b <- lapply(seq_along(a), function(i, cutoff.p) {
    df1 <- a[[i]]
    if (nrow(df1) > 0) {
      df2 <- df1 %>% 
        dplyr::mutate(Library=names(a)[i]) %>%
        dplyr::filter(Adjusted.P.value <= cutoff.p)
    } else {
      df2 <- dummy.df
    }
    if (nrow(df2) > 0) {
      df3 <- dplyr::select(df2, all_of(colms))
    } else {
      df3 <- dummy.df
    }
    df3
  }, cutoff.p=cutoff.p) %>%
    data.table::rbindlist(.) %>% 
    as.data.frame(., stringsAsFactors=FALSE) %>% 
    dplyr::rename(Adj.P.val = Adjusted.P.value) %>% 
    dplyr::distinct(.)
  
  if (sum(complete.cases(b)) > 0) {
    b <- b[complete.cases(b),] %>% 
      dplyr::arrange(., Adj.P.val)
  }
  
  b
}
```


```{r include=FALSE, }
# Code below generates list object.
# Each list item to correspond to one of the gene list items in 'g.lists'.
# Each list item a data frame with enrichment data for multiple enrichr 
# libraries.

enriched.list <- lapply(seq_along(g.lists), function(i, cutoff.adj.p, dbs) {
  genes.i <- g.lists[[i]]
  do.enrichr(gs=genes.i, dbs, cutoff.p=cutoff.adj.p)
}, cutoff.adj.p=cutoff.adj.p, dbs=dbs) %>% 
  set_names(names(g.lists))
out.file.rds <- paste0("./data/", this.script, ".enrichr.RDS")
saveRDS(enriched.list, file=out.file.rds)
```


```{r save RDS, class.output="txt_output"}
cat("File saved:", out.file.rds)
rm(out.file.rds)
```


```{r results="asis"}
x <- paste(utils::capture.output(str(enriched.list)), collapse="<br>\n", sep="")
details::details(x, summary="Show summary of 'enriched.list' object", lang=NULL)
rm(x)
```


### Save in Excel file  
```{r Excel parameters}
xlsx.file <- "Supplemental_Table_2.xlsx"
tab.colors <- c("paleturquoise4", "chartreuse2", "blue", "palevioletred",
                "green", "red", "black", "purple", "yellow")
col.indx <- 1:9
col.widths <- c(50, 50, 12, 12, 12, 12, 254)
```

```{r create Excel workbook}
wb <- openxlsx::createWorkbook()
openxlsx::modifyBaseFont(wb, fontSize = 12, fontName = "Arial")
```

```{r Save analysis info}
info <- data.frame(
  Info=c(
    paste("# Date:", as.character(Sys.time())),
    "# Adjusted p-value cutoff limma: 0.05",
    "# log2FC cutoff limma: log2(1.75)",
    "# Adjusted p-value cutoff enrichment analysis: 1e-06",
    paste("#", R.version.string),
    paste("# Platform:", utils::sessionInfo()$platform),
    paste("# Script:", this.script),
    paste("# Script repository: https://github.com/nohturfft/SuppMat_Milano_et_al_2021_Biorxiv")
  )
)
openxlsx::addWorksheet(wb = wb, sheetName = "Info", gridLines = FALSE, tabColour="yellow")
openxlsx::setColWidths(wb = wb, sheet = "Info", cols = 1, widths = 50)
openxlsx::writeData(wb=wb, sheet="Info", x=info, xy=c("A", 1))
```


```{r Save differentially expressed gene symbol on first sheet}
genes.sheet <- "Genes"
openxlsx::addWorksheet(wb = wb, sheetName = genes.sheet, gridLines = FALSE, tabColour="orange")
openxlsx::setColWidths(wb = wb, sheet = genes.sheet, cols = seq_len(length(g.lists)), widths = 20)
openxlsx::writeData(wb=wb, sheet=genes.sheet,
                    x="Differentially regulated genes (log2FC >= 1.75; pFDR <= 0.05)", xy=c("A", 1))

for (i in seq_along(g.lists)) {
  df <- as.data.frame(g.lists[[i]]) %>% magrittr::set_colnames(names(g.lists)[i])
  openxlsx::writeDataTable(wb=wb, sheet=genes.sheet, x=df, xy=c(LETTERS[i], 3))
}
```


```{r Save GSEA data}
for (i in seq_along(enriched.list)) {
  curr.sheet <- names(enriched.list)[i]
  # Create new worksheet:
  openxlsx::addWorksheet(wb = wb, sheetName = curr.sheet,
                         gridLines = FALSE, tabColour=tab.colors[i])
  # Set column widths:
  openxlsx::setColWidths(wb = wb, sheet = curr.sheet,
                         cols = col.indx, widths = col.widths)
  # Print geneset name
  openxlsx::writeData(wb=wb, sheet=curr.sheet, x=paste("#", curr.sheet), xy=c("A", 1))
  # Print number of input genes
  openxlsx::writeData(wb=wb, sheet=curr.sheet,
                      x=paste("# Number of input genes:", length(g.lists[[curr.sheet]])),
                      xy=c("A", 2))
  # Print number of enrichr hits:
  openxlsx::writeData(wb=wb, sheet=curr.sheet,
                      x=paste("# Number of enrichr hits:", sum(complete.cases(enriched.list[[i]]))),
                      xy=c("A", 3))
  # Print results table:
  if (sum(complete.cases(enriched.list[[i]])) > 0) {
    openxlsx::writeDataTable(wb = wb, sheet = curr.sheet, x = enriched.list[[i]], colNames = TRUE,
                             xy=c("A", 5))
  }
}
```

```{r save info, class.output="txt_output"}
openxlsx::saveWorkbook(wb, xlsx.file, overwrite = TRUE)
cat("Saved:", xlsx.file)
```





## Session info  

<button class="button" onclick="myFunction('DIV_1')">Show/hide session info</button>
<div id="DIV_1" class="div_default_hide">

```{r SESSION INFO DATE, results="asis"}
cat("Date:", format(Sys.time(), "%a %d-%b-%Y %H:%M:%S"), "<br>\n")
```

```{r print_session_info, R.options=list(width=70)}
devtools::session_info()
```
</div>
  
```{js}
var divsToHide = document.getElementsByClassName("div_default_hide");
for(var i = 0; i < divsToHide.length; i++)
{
  divsToHide[i].style.display = 'none';
}
```



